packages/grafana-ui/src/components/Combobox/ComboboxList.tsx

```diff
@@ -0,0 +1,162 @@
+import { cx } from '@emotion/css';
+import { useVirtualizer } from '@tanstack/react-virtual';
+import type { UseComboboxPropGetters } from 'downshift';
+import { useCallback } from 'react';
+
+import { useStyles2 } from '../../themes';
+import { Checkbox } from '../Forms/Checkbox';
+import { ScrollContainer } from '../ScrollContainer/ScrollContainer';
+
+import { AsyncError, NotFoundError } from './MessageRows';
+import { getComboboxStyles, MENU_OPTION_HEIGHT, MENU_OPTION_HEIGHT_DESCRIPTION } from './getComboboxStyles';
+import { ALL_OPTION_VALUE, ComboboxOption } from './types';
+import { isNewGroup } from './utils';
+
+export const VIRTUAL_OVERSCAN_ITEMS = 4;
+
+interface ComboboxListProps<T extends string | number> {
+  options: Array<ComboboxOption<T>>;
+  highlightedIndex: number | null;
+  selectedItems?: Array<ComboboxOption<T>>;
+  scrollRef: React.RefObject<HTMLDivElement>;
+  getItemProps: UseComboboxPropGetters<ComboboxOption<T>>['getItemProps'];
+  enableAllOption?: boolean;
+  isMultiSelect?: boolean;
+  error?: boolean;
+}
+
+export const ComboboxList = <T extends string | number>({
+  options,
+  highlightedIndex,
+  selectedItems = [],
+  scrollRef,
+  getItemProps,
+  enableAllOption,
+  isMultiSelect = false,
+  error = false,
+}: ComboboxListProps<T>) => {
+  const styles = useStyles2(getComboboxStyles);
+
+  const estimateSize = useCallback(
+    (index: number) => {
+      const firstGroupItem = isNewGroup(options[index], index > 0 ? options[index - 1] : undefined);
+      const hasDescription = 'description' in options[index];
+      const hasGroup = 'group' in options[index];
+
+      let itemHeight = MENU_OPTION_HEIGHT;
+      if (hasDescription) {
+        itemHeight = MENU_OPTION_HEIGHT_DESCRIPTION;
+      }
+      if (firstGroupItem && hasGroup) {
+        itemHeight += MENU_OPTION_HEIGHT;
+      }
+      return itemHeight;
+    },
+    [options]
+  );
+
+  const rowVirtualizer = useVirtualizer({
+    count: options.length,
+    getScrollElement: () => scrollRef.current,
+    estimateSize,
+    overscan: VIRTUAL_OVERSCAN_ITEMS,
+  });
+
+  const isOptionSelected = useCallback(
+    (item: ComboboxOption<T>) => selectedItems.some((opt) => opt.value === item.value),
+    [selectedItems]
+  );
+
+  const allItemsSelected = enableAllOption && selectedItems.length === options.length - 1;
+
+  return (
+    <ScrollContainer showScrollIndicators maxHeight="inherit" ref={scrollRef} padding={0.5}>
+      <div style={{ height: rowVirtualizer.getTotalSize() }} className={styles.menuUlContainer}>
+        {rowVirtualizer.getVirtualItems().map((virtualRow, index, allVirtualRows) => {
```

Comment:
Do you need the `index`? 🤔

Reply:
Apart from this comment, it still LGMT 😉

Reply:
no, but we need `allVirtualRows`, the argument after it